### ISSUE TITLE
grub4efi - map of ISO hangs

### ISSUE THREAD

**Issue by steve6375** [chenall/grub4dos#363]
ISO is standard XP Pro SP3 ISO on NTFS volume - tested under VBOX 5.
grub4efi is patched test version from today for issue #361 
![image](https://user-images.githubusercontent.com/1469350/179395713-b89f2baa-48c0-4202-af8e-e17181bfade5.png)
OK with grub4dos 2022-07-17  test version.

**Comment by yaya2007** [comment]
My test here is normal. Probably xp ISO problems. Can you reduce the volume and send it!

**Comment by yaya2007** [comment]
Confirm this xp Can ISO be started through UEFI?

**Comment by steve6375** [comment]
No - it is not UEFI ISO just plain XP - I just tried map command to list Volume ID, etc.
map command should work on any ISO ??

![image](https://user-images.githubusercontent.com/1469350/179479460-0fe79463-20c4-47bc-99f1-26ef6a9f713f.png)
![image](https://user-images.githubusercontent.com/1469350/179479548-6ce84eae-d727-4b0c-8e82-ebf6d9dde20e.png)

Same if XP ISO is  on NTFS or FAT32 partition.
It is standard XP ISO  on an NTFS partition
XP ISO does have a **LOT** of files in i386 folder inside iso.

https://archive.org/download/WinXPProSP3x86/en_windows_xp_professional_with_service_pack_3_x86_cd_vl_x14-73974.iso

**Comment by yaya2007** [comment]
It would be good if the test is normal under grub4dos for BIOS.

This ISO cannot be tested under grub4dos for UEFI.

**Comment by steve6375** [comment]
Yes, works fine in grub4dos. So must be bug in grub4efi when mapping ISO. Fails on both FAT32 and NTFS so it is not a filesystem issue. There are also other problems with map command in grub4efi. Multiple map commands fail, map --unmap does not work, etc.

**Comment by yaya2007** [comment]
First of all, it needs to be clear that the ISO file is started in the UEFI environment, which is conditional, that is, there should be an img image that starts EFI (not the previous img image that starts BIOS). If not, the ISO file cannot be started regardless of whether it is placed in the fat partition or the NTFS partition.

**Comment by memoarfaa** [comment]
@steve6375 
>No - it is not UEFI ISO just plain XP - I just tried map command to list Volume ID, etc.
map command should work on any ISO ??
>
Why you want  to use map command with older xp iso

**Comment by memoarfaa** [comment]
> Yes, works fine in grub4dos. So must be bug in grub4efi when mapping ISO. Fails on both FAT32 and NTFS so it is not a filesystem issue. There are also other problems with map command in grub4efi. Multiple map commands fail, map --unmap does not work, etc.

if its  works fine in grub4dos so its fine

>map command should work on any ISO ??
>
its work in any ISO 
>Multiple map commands fail
>
Multiple map commands work fine

**Comment by steve6375** [comment]
No
Grub4efi map command hangs when used on xp iso. This means there is a bug in code
Try using Map command on 6 different Isos, one after the other. I suggest you test before making unverified statements!

**Comment by wdlkmpx** [comment]
Maybe grub4efi should look for a diferent menu file 
- menu.efi.lst
- menu_efi.lst
- menu.lst

That way the old non-uefi stuff is left out

Or maybe there's a way to hide some menu items in efi mode with a built-in variable like %g4e%

**Comment by memoarfaa** [comment]
@steve6375 

> Grub4efi map command hangs when used on xp iso. This means there is a bug in code
>
there is no bug
 grub4efi map command is not grub2 loopback
in fact you can say it's one parameter of it
it's map ISO9660 filesystem as (fd0)
so never try to map legacy boot ISOs with parameter (hd+) like (hd32) (hd33) etc... or (0x+) like (0xff) (0xa0) etc...

> Try using Map command on 6 different Isos, one after the other.
>
bad testing from you


> I suggest you test before making unverified statements!

Say this to yourself
If you remember me, you wouldn't tell me that
Do you remember who I am? or you still the same person who developed 
[https://rmprepusb.com/](url)  from long time ago 


https://user-images.githubusercontent.com/12494184/180620710-79962641-4d24-45e4-8ced-433faa527303.mp4

**Comment by steve6375** [comment]
I use the the %@uefi% variable
it is not defined in grub4dos, **64** in grub4efi64 and **32** in grub4efi32

So you can use

configfile /my_menu%@uefi%.lst

to go to /my_menu.lst  or /my_menu64.lst or /my_menu32.lst

also, if you rename the utilities in grub4efi \ext folder  you can have common  batch files or menu files which use those utilities

e.g.
fat%@uefi%
hotkey%@uefi%

and the correct one will be called  (e.g. hotkey or hotkey64)

but there are no uefi32 utilities.

**Comment by steve6375** [comment]
for map command issues - see #292

**Comment by steve6375** [comment]
same issue using fd0
![image](https://user-images.githubusercontent.com/1469350/180623135-4a27ce56-2b94-4c86-8493-a56dda9eca0a.png)

command appears to work but actually has not worked.
![image](https://user-images.githubusercontent.com/1469350/180623317-fbe303a2-3df5-4a86-99a1-ea02a2a96937.png)

**Comment by yaya2007** [comment]
have a try
map --no-hook /xpposp3.iso (0xff)

**Comment by memoarfaa** [comment]
hi @yaya2007 
I still try to see how you dial with map command in UEFI.
this parameter is not documented!!!!!.
but I can easily catch it from source code and quickly hit breakpoint to see what its do.

https://user-images.githubusercontent.com/12494184/180630161-02c1aed6-d517-44c5-9936-47e107549eca.mp4

**Comment by memoarfaa** [comment]
@yaya2007 
map --no-hook
The parameter still in development progress 
excuse me if  make some answers in some issues
By the way, I am very thanks to you if you are the same person I think, but the owner of this account encouraged me to learn a lot of programming languages C, C ++, C#, SQL and work in the field of programming as a whole
So far there is no boot loader based on GNU GRUB.
It has this feature that starts the titles from the right to support the right to left language
except grub4dos with setmenu --right - align

![2022-07-25_2-49-11](https://user-images.githubusercontent.com/12494184/180673695-c7d6495d-5734-402b-994d-546b50af6f29.jpg)

I see it's requested here from @a1ive  and he didn't understand it

[https://github.com/a1ive/grub2-filemanager/issues/64#issue-541743712](url)
